utils/src/mesh.rs

```diff
@@ -0,0 +1,113 @@
+use async_trait::async_trait;
+use bincode::{Decode, Encode};
+use if_addrs::Interface;
+use kaboodle::{errors::KaboodleError, Kaboodle};
+
+use std::net::SocketAddr;
+
+/// This type encodes the responsibilities of the resources we are meshing together.
+pub trait KaboodlePeer {
+    fn new(address: SocketAddr, identity: Vec<u8>) -> Self;
+    fn identity(&self) -> Vec<u8>;
+    fn name(&self) -> &str;
+    fn roles(&self) -> Vec<String>;
+}
+
+/// A little wrapper around kaboodle so we can hide the machinery of encoding and decoding.
+/// the identity payload.
+#[async_trait]
+pub trait KaboodleMesh {
+    type A;
+
+    /// Create a new entry for a Kaboodle peer network and add ourselves to the mesh.
+    async fn new(
+        name: &str,
+        roles: Vec<String>,
+        port: u16,
+        interface: Option<Interface>,
+    ) -> Result<Box<Self>, KaboodleError>;
+    /// Get a list of peers. It's the implementer's responsibility to decide if this is fresh or cached somehow.
+    async fn peers(&self) -> Vec<Self::A>;
+    /// Given a specific role, look for a peer that advertises the role.
+    async fn find_role(&self, role: &str) -> Result<Self::A, KaboodleError>;
```

Comment:
Perhaps this will change in a subsequent commit that I haven't looked at yet, but should this be `find_roles` (plural) and return every known peer that has a given role? Because, for example, if we are e.g. looking up a file stored on the mesh, we'll need the instance ids of every storage node to figure out who to ask.